src/dbnode/storage/index/block.go

```diff
@@ -253,30 +258,74 @@ func (b *block) Query(
 		return false, err
 	}
 
-	var (
-		size       = results.Size()
-		brokeEarly = false
-	)
+	size := results.Size()
+	limitedResults := false
 	execCloser := safeCloser{closable: exec}
 	iterCloser := safeCloser{closable: iter}
 
+	// NB(r): iter.Next() when moving between segments will actually
+	// execute the search phase which can take a long time, hence
+	// we only retrieve the results lock when we add a batch of documents
+	// to the results set.
+	batch := b.docsPool.Get()
+	// Use documentArrayPoolCapacity to as max batch to avoid growing outside
+	// the allowed pooled capacity
+	maxBatch := documentArrayPoolCapacity
+
 	defer func() {
 		iterCloser.Close()
 		execCloser.Close()
+		b.docsPool.Put(batch)
 	}()
 
+	// NB(r): This query method only called once per block so is relatively
+	// cheap to declared as a lambda.
+	flushBatch := func() error {
```

Comment:
could you eliminate the allocation entirely if you had a named function / method with the signature:

```golang
func flushBatch(batch []Document, results *ConcurrentResults) ([]Document, error)
```

and then you could use it like:

```golang
batch = append(batch, d)
if len(batch) < maxBatch {
    continue
}
batch, err = flushBatch(batch, results)
if err != nil {
    ...
```

Maybe overkill, but the anonymous func doesn't seem like its closing over many vars